workshops/platformer/README.md

```diff
@@ -0,0 +1,717 @@
+
+```js
+var firstGroundSprite = groundSprites[0];
+if (firstGroundSprite.position.x <= camera.position.x - (width/2)) {
+  groundSprites.remove(firstGroundSprite);
+  firstGroundSprite.position.x = firstGroundSprite.position.x + numGroundSprites*firstGroundSprite.width;
+  groundSprites.add(firstGroundSprite);
+}
+```
+
+Save and refresh live preview. Hm, does something look funny to you? Specifically, how the first ground sprite is just blipping off the left edge instead of smoothly sliding off?
+
+![](img/scrolling_sprites.gif)
+
+This is an issue of not properly offsetting the sprite. Remember that the position of a sprite is at its center, so we must edit our conditional to include this offset. It should now read:
+
+```js
+if (firstGroundSprite.position.x <= camera.position.x - (width/2 + firstGroundSprite.width/2)) {
+```
+
+### Adding Player Controls
+
+We'll define a certain strength our player can jump at the top of the file, underneath gravity:
+
+```js
+var JUMP = -5;
+```
+
+If you're wondering why this value is negative, remember that the y-axis on p5's coordinate system is flipped. Thus, higher numbers are further down, and smaller numbers are further up.
+
+Let's have the up arrow control the jumping. Place this part in `draw()`, right above the line that modifies the player's x-position:
+
+```js
+if (keyDown(UP_ARROW)) {
+}
+```
+
+p5.play has this handy function [`keyDown()`](http://p5play.molleindustria.org/docs/classes/p5.play.html#method-keyDown) that will check if the up arrow key was pressed.
+
+To make the player jump, we'll modify its velocity in the y direction. p5.play has us covered, with this super straightforward `velocity` property, accessible by `player.velocity`. All we have to do is set it to `JUMP`. We'll do this within our conditional:
+
+```js
+if (keyDown(UP_ARROW)) {
+  player.velocity.y = JUMP;
+}
+```
+
+Save and refresh live preview, and give it a try by pressing your up arrow key!
+
+Oops, looks like we forgot to bring the player back down using gravity!
+
+```js
+if (keyDown(UP_ARROW)) {
+  player.velocity.y = JUMP;
+}
+player.velocity.y = player.velocity.y + GRAVITY;
+```
+
+That should do it. The player will definitely come back down now. Save and refresh to check it out.
+
+The player's going down, alright. In fact, it's going through the ground.
+
+We'll have to place a restriction on this. That is, we want to check if the player's y-position is greater than (flipped coordinate system) that of the surface of the ground. If so, we'd like to set the y-velocity back to 0, and ensure that the player rests on the ground.
+
+```js
+if (keyDown(UP_ARROW)) {
+  player.velocity.y = JUMP;
+}
+player.velocity.y = player.velocity.y + GRAVITY;
+
+if (Math.floor(player.position.y) > (height-50)-(player.height/2)) {
+  player.velocity.y = 0;
+  player.position.y = (height-50) - (player.height/2);
```

Comment:
How's it currently handled?

On Sat, May 28, 2016 at 5:41 PM, JESSICA <notifications@github.com> wrote:

> In workshops/platformer/README.md
> <https://github.com/hackclub/hackclub/pull/827#discussion_r64996154>:
>
> > +
> > +That should do it. The player will definitely come back down now. Save and refresh to check it out.
> > +
> > +The player's going down, alright. In fact, it's going through the ground.
> > +
> > +We'll have to place a restriction on this. That is, we want to check if the player's y-position is greater than (flipped coordinate system) that of the surface of the ground. If so, we'd like to set the y-velocity back to 0, and ensure that the player rests on the ground.
> > +
> > +```js
> > +if (keyDown(UP_ARROW)) {
> > +  player.velocity.y = JUMP;
> > +}
> > +player.velocity.y = player.velocity.y + GRAVITY;
> > +
> > +if (Math.floor(player.position.y) > (height-50)-(player.height/2)) {
> > +  player.velocity.y = 0;
> > +  player.position.y = (height-50) - (player.height/2);
>
> Question: should we actually have the player object collide with the
> ground objects or something? and do if (player.overlap(ground)) { put the
> player on top of the ground } Now that I've written this comment, I feel
> like yes. @zachlatta <https://github.com/zachlatta>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/hackclub/hackclub/pull/827/files/3e865b06d2d6db1994503b8ebce070d1416d038f#r64996154>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA8j-KYCNovj12uGlII-2-mJf8WT2rR8ks5qGODMgaJpZM4Iepj8>
> .
>



-- 
Zach Latta
https://zachlatta.com <http://zachlatta.com>